modules/calib3d/include/opencv2/calib3d.hpp

```diff
@@ -3270,14 +3270,1282 @@ optimization. It stays at the center or at a different location specified when C
+
+/** @brief Solves the PnP problem to compute the 3D pose of an object, using a set of 3D-to-2D point correspondences and a perspective camera.
+
+    @param opoints Set of Object points. This must be a 1xN 3-channel or Nx1 3-channel matrix (double) where n is the number of points.
+    opoints are defined in object coordinates.
+    @param ipoints Set of Image points. This must be a 1xN 2-channel or Nx1 2-channel matrix (double) where n is the number of points.
+    ipoints are defined in normalized pixel coordinates. That is, they are defined by using a canonical intrinsic calibration matrix of K = eye(3).
+    You can convert a set of points defined in pixels to normalized pixel coordinates using @ref undistortPoints. This is convenient because it
+    eliminates the need for supplying the camera's intrinsics (calibration matrix and distortion parameters)
+    @param rvecs Vector of output rotation vectors (double) (see @ref Rodrigues ) that, together with tvecs, brings points from
+    the model coordinate system to the camera coordinate system.
+    @param tvecs Vector of output translation vectors (double).
+    @param sortOnReprojectionError Sorts the pose solutions with lowest reprojection error first
+    @returns if sortOnReprojectionError then returns the sorted reprojection errors corresponding to each pose solution (descending order)
+    otherwise returns an empty vector
+     */
+    std::vector<double> solveProblem(InputArray opoints, InputArray ipoints, CV_OUT std::vector<Mat> & rvecs,
+                                     CV_OUT std::vector<Mat> & tvecs, bool sortOnReprojectionError = true) const;
+
+
+/** @brief Checks if the number of object points n can be handled by the algorithm.
+    This returns n >= minPointNumber() && n <= maxPointNumber()
+    In the case that the derived solver cannot solve n for values between  minPointNumber() and maxPointNumber()
+    this should be overriden to specify it.
+
+    @param opoints Set of Object points. This must be a 1xN 3-channel or Nx1 3-channel matrix (double) where n is the number of points.
+    opoints are defined in object coordinates.
+    @returns true if number can be handled otherwise false
+     */
+    CV_WRAP virtual bool checkNumberOfPoints(InputArray opoints) const;
+
+/** @brief Gives the minimal number of object points handled by the algorithm
+
+    @returns Minimal number of points supported by algorithm
+     */
+    CV_WRAP virtual int minPointNumber() const = 0;
+
+/** @brief Gives the Maximal number of object points handled by the algorithm. If there is no maximum this returns -1.
+
+    @returns Maximal number of points supported by algorithm. If no maximum, returns -1
+     */
+    CV_WRAP virtual int maxPointNumber() const = 0;
+
+/** @brief Indicates if the algorithm requires the object points to be 3D (i.e. not co-planar)
+
+    @returns true if required, false otherwise
+     */
+    CV_WRAP virtual bool requires3DObject() const = 0;
```

Comment:
how about using a single `int getCaveats() const` function returning a mask like `PNP_CAVEAT_PLANAR | PNP_CAVEAT_TAG` instead of the three functions below?

The same mask could be also returned from `validProblem` to signal which check failed.